### DISCUSSION TITLE
WinUI3 VS2022 .Net6?

### DISCUSSION THREAD

**Discussion by Tide** [microsoft/CsWin32#514] [General]
I am trying to use CSWin32 in a .Net6 WinUI3 project in VS2022.  I installed the v0.1.635-beta without issue.   However, When I try to add using Windows.Win32; I am getting errors.  

Is CsWin32 supposed to work in my scenario?

**Comment by AArnott** [comment]
Yes. What errors do you see?

**Comment by Tide** [comment]
@AArnott  I installed the package via command line.  That goes fine and you can see it in the packages section of the solution.  However, when you try to add the using statement it is just red and can't find "Windows.CsWin32".  If I type it by hand there is not CsWin32 option available under "Windows".  If I download some app that are .Net5 and reference CsWin32 they seem to work just fine.  Adding the using statement is as far as I can get.

Thanks for you help,
Tide

**Reply by Tide** [reply]
Later today I will try to create a test repo.  However, to reproduce the issue just create a new WinUi3 .Net6 solution.  Then add the latest CsWin32 via command line.  Add the using statement to the top of a class and you will see the error.

**Comment by Tide** [comment]
@AArnott  I added a sample repo to show the issue. 
https://github.com/Tide/CsWin32Test

If you try to install CsWin32 via graphical Nuget you will not find it available for WinUI3 .Net6 apps.  You can only install it via command line. `Install-Package Microsoft.Windows.CsWin32 -Version 0.1.635-beta`.  Not that in itself is the issue but I am guessing it is supposed to show up there so I thought I would mention that as well.  

After you look at it if you want me to create an issue for it I can.  I did not do one initially because I thought it may be on my side.

**Reply by AArnott** [reply]
The package shows up in the Nuget Package Manager if you check the "Include prereleases" checkbox:
![image](https://user-images.githubusercontent.com/3548/158717450-664f50b1-1398-4e68-b27a-1ed72fdafaed.png)

**Comment by Tide** [comment]
Here is a picture showing the package installed and the using not working.

![Untitled](https://user-images.githubusercontent.com/237330/158073414-c8288006-84e2-42f5-9bb6-cbb296b9b053.jpg)

**Reply by AArnott** [reply]
Your [NativeMethods.txt file](https://github.com/Tide/CsWin32Test/blob/main/CsWin32Test/CsWin32Test/NativeMethods.txt) is empty. CsWin32 only generates the APIs you ask for. Since you're not asking for any, nothing gets generated into that namespace.
Please review the readme file.

**Comment by Tide** [comment]
@AArnott You are correct that was the issue.  I am guessing source is being generated behind the scenes as I am learning.  It is just odd to go forward even though the using was red.  As I first felt it was me and not this library.  

Thank you for your help on this!

**Comment by GuildOfCalamity** [comment]
I just ran into this same issue, Intellisense does not honor the using/namespace.
I think the NuGet should be updated with a popup tip indicating that a **NativeMethods.txt** file is missing/required in the event that someone misses that in the ReadMe.

**Reply by AArnott** [reply]
As a source generator, CsWin32's options for a 'popup' don't really exist. But we could generate a build error or warning. That would only appear in an actual build though, since C# doesn't manifest source-generator produced diagnostics in the Error List, sadly. And I imagine if folks are trying to figure out why a namespace isn't showing up in completion lists, they probably aren't going to Build and check the output window. At least so far the customer feedback has suggested this to be the case at least sometimes. 
Do you have specific ideas for how this could be reported to the user for us to evaluate?

**Reply by GuildOfCalamity** [reply]
Just as you deliver a **readme**, would it be possible to deliver a boilerplate **NativeMethods.txt** which contains something so that the namespace resolves? Since this package operates differently from most, I think an extra step is required.

**Reply by AArnott** [reply]
That would make sense, but the package doesn't add the readme to your project. NuGet's IDE integration merely opens the readme from the package up for your review. There is no way that I know to add source files to your project that you can edit. The package can only add source files that are readonly to you (because they come from the package, or the source generator). 
With the old style nuget (packages.config), there was an install.ps1 script that traditionally could be used for this purpose. But that's a long-gone era.

But hey, if you see the readme appear, would it help if the required steps were made more prominent in the readme?

**Reply by GuildOfCalamity** [reply]
Yeah, I guess that's a tricky one. More attention to the part about creating the **NativeMethods.txt** file would help for sure.
At any rate, I love the work you've done... I just don't want others to get frustrated with it (like I did). 
Maybe you've already explained this, but why exactly does the namespace not become recognized/resolved in VS without adding the txt file?

**Reply by AArnott** [reply]
Without the txt file, no APIs are generated, and thus no namespace around them.
I do appreciate the feedback. I'm just not sure what to do about it. Although it sounds like adjusting the readme may help.
